sdk/core/Azure.Core.TestFramework/src/TestProxy.cs

```diff
@@ -0,0 +1,230 @@
+                        // output to console in case another error in the test causes the exception to not be propagated
+                        TestContext.Progress.WriteLine(error);
+                        _errorBuffer.AppendLine(error);
+                    }
+                });
+
+            int lines = 0;
+            while ((_proxyPortHttp == null || _proxyPortHttps == null) && lines++ < 20)
+            {
+                string outputLine = _testProxyProcess.StandardOutput.ReadLine();
+                // useful for debugging
+                TestContext.Progress.WriteLine(outputLine);
+
+                if (ProxyPortHttp == null && TryParsePort(outputLine, "http", out _proxyPortHttp))
+                {
+                    continue;
+                }
+
+                if (_proxyPortHttps == null && TryParsePort(outputLine, "https", out _proxyPortHttps))
+                {
+                    continue;
+                }
+            }
+
+            if (_proxyPortHttp == null || _proxyPortHttps == null)
+            {
+                throw new InvalidOperationException("Failed to start the test proxy.");
+            }
+
+            // we need to use https when talking to test proxy admin endpoint so that we can establish the connection before any
+            // test related traffic happens which would send a Connection header for the first request. This can be switched to HTTP
+            // once https://github.com/Azure/azure-sdk-tools/issues/2303 is fixed
+            var handler = new HttpClientHandler
+            {
+                ServerCertificateCustomValidationCallback = (_, certificate, _, _) => certificate.Issuer == DevCertIssuer
+            };
+            var options = new TestProxyClientOptions
+            {
+                Transport = new HttpClientTransport(handler)
+            };
+            Client = new TestProxyRestClient(
+                new ClientDiagnostics(options),
+                HttpPipelineBuilder.Build(options),
+                new Uri($"https://{IpAddress}:{_proxyPortHttps}"));
+
+            // For some reason draining the standard output stream is necessary to keep the test-proxy process healthy. Otherwise requests
+            // start timing out. This only seems to happen when not specifying a port.
+            _ = Task.Run(
+                () =>
+                {
+                    while (!_testProxyProcess.HasExited && !_testProxyProcess.StandardOutput.EndOfStream)
+                    {
+                        _testProxyProcess.StandardOutput.ReadLine();
+                    }
+                });
+        }
+
+        public static TestProxy Start()
+        {
+            return new TestProxy(typeof(TestProxy)
+                .Assembly
+                .GetCustomAttributes<AssemblyMetadataAttribute>()
+                .Single(a => a.Key == "TestProxyPath")
+                .Value);
+        }
+
+        public void Stop()
+        {
+            _testProxyProcess?.Kill();
```

Comment:
IIRC, this will throw if already killed. Perhaps nullify it after this call?

Reply:
This would only be called once so it should be okay - if it is called more than once there is a bug that we should fix.